File: perf/benchmarks/apps/histogram/histogram.cpp
// histogram — bucket-counts of LCG ints application benchmark (C++).
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <vector>

constexpr int K = 1024;

int main(int argc, char** argv) {
    int64_t n = (argc > 1) ? std::atoll(argv[1]) : 5000000;
    auto t0 = std::chrono::steady_clock::now();
    std::vector<int64_t> buckets(K, 0);
    int64_t seed = 1;
    for (int64_t i = 0; i < n; ++i) {
        seed = (seed * 1103515245LL + 12345LL) & 0x7FFFFFFFLL;
        buckets[(seed >> 8) & (K - 1)] += 1;
    }
    int64_t cs = 0;
    for (auto v : buckets) cs ^= v;
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::printf("CHECKSUM:%lld\n", static_cast<long long>(cs));
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}

File: perf/benchmarks/apps/histogram/histogram.rs
// histogram — bucket-counts of LCG ints application benchmark (Rust).
use std::env;
use std::time::Instant;

const K: usize = 1024;

fn main() {
    let args: Vec<String> = env::args().collect();
    let n: i64 = if args.len() > 1 { args[1].parse().unwrap_or(5_000_000) } else { 5_000_000 };
    let t0 = Instant::now();
    let mut buckets: Vec<i64> = vec![0; K];
    let mut seed: i64 = 1;
    for _ in 0..n {
        seed = (seed.wrapping_mul(1103515245).wrapping_add(12345)) & 0x7FFF_FFFF;
        buckets[((seed >> 8) as usize) & (K - 1)] += 1;
    }
    let mut cs: i64 = 0;
    for v in &buckets { cs ^= v; }
    let ms = t0.elapsed().as_nanos() as f64 / 1_000_000.0;
    println!("CHECKSUM:{}", cs);
    println!("ELAPSED_MS:{:.6}", ms);
}
